src/core/thread/mle.cpp

```diff
@@ -4743,5 +4773,44 @@ void Mle::DelayedResponseMetadata::RemoveFrom(Message &aMessage) const
     SuccessOrAssert(aMessage.SetLength(aMessage.GetLength() - sizeof(*this)));
 }
 
+Error Mle::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    OT_ASSERT(IsChild());
+
+    VerifyOrExit(!IsDetachingGracefully(), error = kErrorBusy);
+
+    mDetachGracefullyCallback = aCallback;
+    mDetachGracefullyContext  = aContext;
+
+    mDetachGracefullyTimer.Start(kDetachGracefullyTimeout);
+
+    IgnoreError(SendChildUpdateRequest(/*aTimeout=*/0));
+
+exit:
+    return error;
```

Comment:
I think this method here in `Mle` and the `MleRouter` one are very similar and share a good chunk of code (checking if we are already detaching, setting callback, scheduling timer, ect.) and differ in the action taken (send child update or release address based of current role)

How about we combine them into one method?

We can have it be defined in `Mle` and have `#if OPENTHREAD_FTD` around the parts that are FTD specific.

Reply:
Done.